packages/loader/container-loader/src/deltaManager.ts

```diff
@@ -60,9 +60,11 @@ const ImmediateNoOpResponse = "";
 // eslint-disable-next-line @typescript-eslint/no-unsafe-return
 const getRetryDelayFromError = (error: any): number | undefined => error?.retryAfterSeconds;
 
-function getNackReconnectInfo(nackContent: INackContent) {
+function getNackReconnectInfo(nackContent: INackContent, mode: ConnectionMode) {
     const reason = `Nack: ${nackContent.message}`;
-    const canRetry = ![403, 429].includes(nackContent.code);
+    // always retry/reconnect on Nack if connection mode is read, as first
+    // write on a read connection should always nack first, then re-connect as write
+    const canRetry = mode === "read" ? true : ![403, 429].includes(nackContent.code);
```

Comment:
This looks to be an implementation designed for solve a problem that exists at ContainerRuntime level.
Would we be better of finding a solution at that layer instead?
How things will behave if client has only read access to file, and every connection attempt (read or write) always returns read connection?

The reason I'm not sure that's the right fix is that we already have too many things in DeltaManager that are solutions for problems at Container or ContainerRuntime (like reconnecting at "write" mode if there are any non-acked ops).

Reply:
The concept of read is so overloaded right now from read only connection, to read only permission, and the existing implementations doesn't do a good job of differentiating the different paths. I'd be all for a rewrite, but that seems a bit out of scope here.  From the bug it sounds like this is an active issue if it's not, i'm happy to open a new one for do a more throughout rewrite. If it's an active issue we can also add a check for read only permission here, and still open the folllow up.

Reply:
For this issue specifically, if a user only has readonly access to the file, the reconnect will fail, and should have been blocked by the app from making modifications (this may be another bug), which seems unrelated to summaries. A user with readonly permission can never summarize, as they will never get a write connection, 

Are you suggesting the container runtime should try to spawn the summarizer even if its connection is readonly? That would require completely redoing leader election for the summarizer, which also seems out of scope here.

Reply:
No, reconnect does not have to fail. Server will downgrade request and return read connection. This will result in us realizing user does not have write permissions and raising "readonly" event ( believe that's how code works, though we double-check that with claims, I think).

Nack contains enough information to differentiate the need to upgrade to write (and possibly ability to do so), vs. nack due to too many ops and not able to do so. All I'm saying is that we chose to connect to write when we know it will fail, only to jump start summarizer. Can this be done without it? 
Can container runtime listen for this same nack error (I believe we propagate it through dispose() flow) and if it's 10K ops, and there are no write clients in quorum, jump start summarizer?
It will have exactly same race condition as proposed workflow (by time potential summarizer client loads and connects it maybe not oldest client, and thus not summarizer - we do not have this race in normal case, but we do in this case where main client can maintain write connection).

Reply:
The issue is we elect summarizer based on main client, not summarizer client, so having a read-only main client will materially change the leader election and introduce a new race, which doesn't exist today.  

For this specific issue, we only nack the main client after they attempted to write, so they have outstanding ops, which shouldn't happen if they don't have permissions (but could due to bugs). From my understanding if they have write permissions, they should be able to successfully connect write, it's just the ops that will get nack. Since they can connect write, there no reason to not just use the normal summarizer election flow. Once they connect we will spawn the summary container, which should live longer than the main client to complete summarization.

Reply:
I fully agree with the impulse to move much of this handling out of the loader/delta manager, I'm just not sure this specific issue is the right place to start. I'd rather tackle that problem independently rather than potentially shoe horing this issue in.

Reply:
@arinwt should confirm (or deny), but for last "last" summarizer we do not use main client for calculation, as it may not exists (including host closing it). This is true here as well, as we get nack because main client is attempting to send an op - it will be ale to get write connection, but it will be nacked right away (and close connection) because we will resend that same op and fail server's 10K limit. So "last" summarizer will be booting without main container on the wire, and it will re-access if it still qualifies to be a summarizer after it connects to delta connection.

So I believe one way or another, starting summarizer client is racing to become a summarizer, without main client in the picture (it's gone). The only difference is the mechanism that starts summarizing client - in proposed solution it starts due to very short lived "write" connection of main client. I propose we look into starting it on 10K op nack directly (i.e. this logic is not in deltaManager layer, as much as possible, but right signals are propagated to ContanerRuntime to be able to do right things).

Reply:
@arinwt - would love your opinion on that. I think that there is a day of work max to ensure that we have unique errorType for 10K ops nack and couple lines of code in ContainerRuntime.dispose to check for this error, check quorum to be empty and fire summarizer.

Reply:
Vlad is correct, the current behavior is that summarizer client can run without its parent connected, but it never _starts_ without a connected write client as its parent.

So the basic flow is this:
1. the parent client connects as write (even if only momentarily)
2. then parent spawns summarizer client
3. then parent disconnects maybe even before summarize client connects
4. then summarizer sees parent has disconnected and tries to send out one last summary before closing

So Tony is right that it is adding a bit of a new paradigm: _starting_ summarizer client without being connected as write.  Part of the benefit of being connected as write client before spawning summarizer is that the client knows it is in the collab window.

But yes, I think code-wise, we could have SummaryManager attempt to spawn a summarizer in the scenario Vlad mentioned (do we also want to check it is only client in Audience? or just let them all attempt?). I don't think the consequences are that bad in the "bad" case, which is that we end up with multiple simultaneous summarizers and some get nacked.  Just need to make sure everything is kosher in the SummaryManager state machine.

Reply:
Also I think in this case, we want the spawned summarizer to close itself after finishing its summary, does that seem reasonable?  So it just runs it and immediately stops it. That way we don't end up with these extra summarizer clients lingering around after the doc has stabilized and write clients are joining.  Although write clients will not try to spawn a new summarizer client if one is already on the wire, the alternative would be fine too, the problem is I think the state machine for read clients is a little harder to manage, but maybe I'm wrong.